crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity.rs

```diff
@@ -1,157 +1,223 @@
+    let chain_filter_limit_conditions_until_hashes_max_seq =
+        get_chain_filter_limit_conditions_until_hashes_max_seq(scratch, filter.clone());
+
+    // Until hash Action seq must be less than or equal to ChainFilter chain_top Action seq
+    if let Some(until_action_seq) = chain_filter_limit_conditions_until_hashes_max_seq {
+        if until_action_seq > filter_chain_top_action_seq {
+            return Err(StateQueryError::InvalidInput("The largest ChainFilter until hash Action seq must be less than or equal to the ChainFilter chain_top action seq.".to_string()));
         }
-        let result = by_hash_statement
-                .query_row(named_params! {":hash": hash, ":author": author, ":activity": ChainOpType::RegisterAgentActivity}, |row| {
-                    row.get(0)
-                })
-                .optional()?;
-        Ok(result)
-    };
+    }
 
-    // Get the oldest action in a given time period.
-    let mut by_ts_statement = txn.prepare(TS_TO_SEQ)?;
-    let get_seq_from_ts = |until: Timestamp| {
-        // Search in DB
-        let db_result = by_ts_statement
-          .query_row(named_params! {":until": until.as_micros(), ":author": author, ":activity": ChainOpType::RegisterAgentActivity}, |row| {
-              row.get(0)
-          })
-          .optional()?;
-        // Search in scratch
-        let scratch_result = scratch.and_then(|s| {
-            s.actions()
-                .filter_map(|a| {
-                    if a.get_timestamp() >= until {
-                        Some(a.seq())
-                    } else {
-                        None
-                    }
-                })
-                .min()
-        });
-        // Return lowest of the two
-        Ok([db_result, scratch_result].into_iter().flatten().min())
-    };
+    // Get the agent activity, filtered by the chain top, author, 3 optional lower-bounds, and optional limit size.
+    let activity = scratch
+        .actions()
+        .filter(|a| {
+            let action = a.action();
+            let is_author = action.author() == author;
+            let is_lte_chain_top = action.action_seq() <= filter_chain_top_action_seq;
+
+            let mut is_gte_until_timestamp = true;
+            if let Some(until_timestamp) = filter.get_until_timestamp() {
+                is_gte_until_timestamp = a.action().timestamp() >= until_timestamp;
+            }
+
+            let mut is_gte_max_until_hash_seq = true;
+            if let Some(until_action) = chain_filter_limit_conditions_until_hashes_max_seq {
+                is_gte_max_until_hash_seq = a.action().action_seq() >= until_action;
+            }
+
+            is_author && is_lte_chain_top && is_gte_until_timestamp && is_gte_max_until_hash_seq
+        })
+        .map(|action| RegisterAgentActivity {
+            action: action.clone(),
+            // TODO: Implement getting the cached entries.
+            cached_entry: None,
```

Comment:
I think that whole feature should go, it's an early optimization with no evidence that I've seen for why we need it.